framework/data/ActiveDataProvider.php

```diff
@@ -93,14 +94,51 @@ public function init()
     }
 
     /**
-     * {@inheritdoc}
+     * Creates a wrapper of [[query]] that allows adding limit and order.
+     * @return QueryInterface
+     * @throws InvalidConfigException
      */
-    protected function prepareModels()
+    protected function createQueryWrapper(): QueryInterface
     {
         if (!$this->query instanceof QueryInterface) {
             throw new InvalidConfigException('The "query" property must be an instance of a class that implements the QueryInterface e.g. yii\db\Query or its subclasses.');
         }
-        $query = clone $this->query;
+        if (!$this->query instanceof Query || empty($this->query->union)) {
+            return clone $this->query;
```

Comment:
Do we need a clone here?

Reply:
Tests pass successfully without clone. I think it can be removed.

Reply:
But this kind of test will fail without clone.
```php
$provider = new ActiveDataProvider([
    'query' => Item::find(),
]);
$pagination = $provider->getPagination();
$pagination->pageSize = 2;
$provider->prepare();
$this->assertCount(2, $provider->getModels());

$provider->setPagination(false);
$provider->prepare(true);
$this->assertCount(5, $provider->getModels());
```

Reply:
Makes sense,